test/xds/xds_client_ignore_resource_deletion_test.go

```diff
@@ -0,0 +1,474 @@
+						ClusterSpecifier: &routepb.RouteAction_Cluster{Cluster: cdsName1},
+					}},
+				},
+				{
+					Match: &routepb.RouteMatch{PathSpecifier: &routepb.RouteMatch_Prefix{Prefix: "/grpc.testing.TestService/UnaryCall"}},
+					Action: &routepb.Route_Route{Route: &routepb.RouteAction{
+						ClusterSpecifier: &routepb.RouteAction_Cluster{Cluster: cdsName2},
+					}},
+				},
+			},
+		}},
+	}
+)
+
+// This test runs subtest each for a Listener resource and a Cluster resource deletion
+// in the response from the server for the following cases:
+//   - testResourceDeletionIgnored: When ignore_resource_deletion is set, the
+//     xDSClient should not delete the resource.
+//   - testResourceDeletionNotIgnored: When ignore_resource_deletion is unset,
+//     the xDSClient should delete the resource.
+//
+// Resource deletion is only applicable to Listener and Cluster resources.
+func (s) TestIgnoreResourceDeletionOnClient(t *testing.T) {
+	port1, cleanup := startTestService(t, nil)
+	t.Cleanup(cleanup)
+
+	port2, cleanup := startTestService(t, nil)
+	t.Cleanup(cleanup)
+
+	initialResourceOnServer := e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: []*listenerpb.Listener{e2e.DefaultClientListener(serviceName, rdsName)},
+		Routes:    []*routepb.RouteConfiguration{defaultRouteConfigWithTwoRoutes},
+		Clusters: []*clusterpb.Cluster{
+			defaultClientCluster(cdsName1, edsName1),
+			defaultClientCluster(cdsName2, edsName2),
+		},
+		Endpoints: []*endpointpb.ClusterLoadAssignment{
+			e2e.DefaultEndpoint(edsName1, "localhost", []uint32{port1}),
+			e2e.DefaultEndpoint(edsName2, "localhost", []uint32{port2}),
+		},
+		SkipValidation: true,
+	}
+
+	tests := []struct {
+		name           string
+		updateResource func(r *e2e.UpdateOptions)
+	}{
+		{
+			name: "listener",
+			updateResource: func(r *e2e.UpdateOptions) {
+				r.Listeners = nil
+			},
+		},
+		{
+			name: "cluster",
+			updateResource: func(r *e2e.UpdateOptions) {
+				r.Clusters = nil
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%s resource deletion ignored", test.name), func(t *testing.T) {
+			testResourceDeletionIgnored(t, initialResourceOnServer, test.updateResource)
+		})
+		t.Run(fmt.Sprintf("%s resource deletion not ignored", test.name), func(t *testing.T) {
+			testResourceDeletionNotIgnored(t, initialResourceOnServer, test.updateResource)
+		})
+	}
+}
+
+// This subtest tests the scenario where the bootstrap config has "ignore_resource_deletion"
+// set in "server_features" field. This subtest verifies that the resource was
+// not deleted by the xDSClient when a resource is missing the xDS response and
+// RPCs continue to succeed.
+func testResourceDeletionIgnored(t *testing.T, resources e2e.UpdateOptions, u func(r *e2e.UpdateOptions)) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	t.Cleanup(cancel)
+	ms := startManagementServer(t)
+	bootstrapContent := generateBootstrapContents(t, ms.Address, true)
```

Comment:
Nit: s/bootstrapContent/bs/
And move the call to `generateCustomResolver` to be right below this line (because technically that is also part of setup).

Comment applies to other tests where the same pattern is used.

Reply:
done.